spec/app/ics-028-cross-chain-validation/technical_specification.md

```diff
@@ -0,0 +1,1088 @@
+```typescript
+// PCF: Provider Chain Function
+// implements governance proposal Handler 
+function CreateConsumerChainProposal(p: CreateConsumerChainProposal) {
+  if currentTimestamp() > p.spawnTime {
+    // get UnbondingPeriod from provider Staking module
+    unbondingTime = stakingKeeper.UnbondingTime()
+
+    // create client state as defined in ICS 2
+    clientState = ClientState{
+      chainId: p.chainId,
+      trustLevel: DefaultTrustLevel, // 1/3
+      trustingPeriod: unbondingTime/2,
+      unbondingPeriod: unbondingTime,
+      latestHeight: p.initialHeight,
+    }
+
+    // create consensus state as defined in ICS 2;
+    // SentinelRoot is used as a stand-in root value for 
+    // the consensus state set at the upgrade height
+    consensusState = ConsensusState{
+      timestamp: currentTimestamp(),
+      root: SentinelRoot,
+      nextValidatorsHash: BlockHeader().NextValidatorsHash,
+    }
+
+    // create consumer chain client and store it
+    clientId = clientKeeper.CreateClient(clientState, consensusState)
+    consumerClient[p.chainId] = clientId
+  }
+  else {
+    // store the client as a pending client
+    pendingClient[(p.spawnTime, p.chainId)] = p.initialHeight
+  }
+}
+```
+- Initiator: 
+  - `EndBlock()` method of Governance module.
+- Expected precondition: 
+  - A governance proposal with `CreateConsumerChainProposal` as content has passed (i.e., it got the necessary votes). 
+- Expected postcondition: 
+  - If the spawn time has already passed,
+    - `UnbondingPeriod` is retrieved from the provider Staking module;
+    - a client state is created;
+    - a consensus state is created;
+    - a client of the consumer chain is created and the client ID is added to `consumerClient`.
+  - Otherwise, the client is stored in `pendingClient` as a pending client.
+- Error condition:
+  - None.
+
+> **Note:** Creating a client of a remote chain requires a `ClientState` and a `ConsensusState` (as defined in [ICS 2](../../core/ics-002-client-semantics)). For Tendermint clients, creating a `ConsensusState` requires setting a validator set of the remote chain (see [ICS 7](../../client/ics-007-tendermint-client/README.md#consensus-state)). The provider chain uses the fact that the validator set of the consumer chain is the same as its own validator set. The rest of information to create a `ClientState` it receives through a governance proposal.
+
+<!-- omit in toc -->
+#### **[CCV-PCF-COINIT.1]**
+```typescript
+// PCF: Provider Chain Function
+// implements the ICS26 interface
+function onChanOpenInit(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) {
+    // the channel handshake MUST be initiated by consumer chain
+    abortTransactionUnless(FALSE)
+}
+```
+- Initiator: 
+  - The IBC module on the provider chain.
```

Comment:
Indeed, that was the reason for having the IBC module as the initiator. I think the purpose of the "Initiator" is a bit unclear; it may be because of the name. In my view, there should be two different concepts that need to be specify for every function:
- The module that calls the function. This enables us to reason about the external dependencies (e.g., if `onChanOpenInit` is called by the IBC module, then any changes to the IBC module will affect the logic in this function).
- The event that triggered the function to be called. This enables us to reason about the correctness of the protocol (e.g., `onChanOpenInit` is called as a result of a relayer submitting a `MsgChanOpenInit`). 

We could name the first one "Caller", and the second one "Trigger". What do you think?